torch/csrc/distributed/c10d/reducer.cpp

```diff
@@ -803,15 +797,12 @@ void Reducer::mark_variable_ready(size_t variable_index) {
   // TODO(@pietern): Make this work for both CPU/CUDA tensors.
   // When using CPU tensors we don't need to do this.
   // // Record event so that we can wait for all of them.
-  // auto& event = replica.events[bucket_index.intra_bucket_index];
+  // auto& event = bucket.events[bucket_index.intra_bucket_index];
   // event.record();
```

Comment:
(can be done in followup PR) this comment seems still relevant. The autograd hook will fire on the same stream used by that function during forward, which might be different from the allreduce stream. I didn't we record that stream anywhere in reducer. 

https://github.com/pytorch/pytorch/blob/ec3a5ca6d3f7762922effc3d44132a7723bf8e16/torch/csrc/distributed/c10d/reducer.cpp#L376-L380

@zhaojuanmao @rohan-varma would I be correct that we should add this to guarantee correctness?

If we still need to make it work for CPU, today, we can use c10 events, which will work seamlessly for both CPU and GPU

Reply:
See how `ProcessGroupGloo` used `c10::event` and `c10::stream`. They basically become noop for CPU devices. 

https://github.com/pytorch/pytorch/blob/ec3a5ca6d3f7762922effc3d44132a7723bf8e16/torch/csrc/distributed/c10d/ProcessGroupGloo.cpp#L330-L364

Reply:
@mrshenli I did have a fix a while back for DDP under non-default stream: https://github.com/pytorch/pytorch/pull/40115/files

In this fix, we have a test where forward uses non-default stream and autograd engine restores these streams during backward. The fix for DDP was to sync with the current stream as opposed to default stream. Seems like that takes care of it, is there anything else we need to do?

Reply:
Ah I see, yep, that should work. Thanks! In that case, we can remove Pieter's comments?